clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/codecs/OperationCodecFactory.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ehcache.clustered.client.internal.store.operations.codecs;
+
+import org.ehcache.spi.serialization.Serializer;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class OperationCodecFactory<K, V> {
+
+  private final Serializer<K> keySerializer;
+  private final Serializer<V> valueSerializer;
+
+  private final Map<OperationCode, OperationCodec<K>> operationCodeVsCodecMap;
+
+  public OperationCodecFactory(final Serializer<K> keySerializer, final Serializer<V> valueSerializer) {
+    this.keySerializer = keySerializer;
+    this.valueSerializer = valueSerializer;
+    operationCodeVsCodecMap = populateCodecMap();
+  }
+
+  private Map<OperationCode, OperationCodec<K>> populateCodecMap() {
+    Map<OperationCode, OperationCodec<K>> map = new HashMap<OperationCode, OperationCodec<K>>();
+    for (OperationCode code : OperationCode.values()) {
+      operationCodeVsCodecMap.put(code, getNewCodec(code));
+    }
+    return map;
+  }
+
+  public OperationCodec<K> getOperationCodec(OperationCode code) {
+    return operationCodeVsCodecMap.get(code);
+  }
+
+  private OperationCodec<K> getNewCodec(OperationCode code) {
```

Comment:
Well things could get a bit weird if I embark on that path. Or may be I'm not understanding your suggestion properly: 

If I add a method at the enum level returning an `AbstractOperationCodecFactory` which has a structure something like this: 

```java
abstract class AbstractOperationCodecFactory<K, V> {
  public abstract OperationCodec<K> getOperationCodec(Serializer<K> keySerializer, Serializer<V> valueSerializer);
}
```
and a concrete implementation, like for example `PutOperationCodecFactory` with a structure like this:

```java
public class PutOperationCodecFactory<K, V> extends AbstractOperationCodecFactory<K, V> {

  OperationCodec<K> getOperationCodec(Serializer<K> keySerializer, Serializer<V> valueSerializer) {
    return new PutOperationCodec<K, V>(keySerializer, valueSerializer);
  }
}
```
If we take this path, then for every encode method call, we'll end up calling:

Enum.getOperationCodec(keySer, valueSer).encode(operation)

Every _getOperationCodec_ call is gonna return a new instance of the codec with the provided serializer pairs and finally we end up with too many `OperationCodec` instances. So I'm not sure if this is the right thing to do.

Now if we decide to cache the instance of codec on the first _getOperationCodec_ call and then decide to reuse it on subsequent calls( singleton like approach), we'll end up with an enum that'll be tied to a single pair of serializers which in-turn ties the enum to a single cache. I don't think that's what we want either.

These were the two approaches that came to my mind. Did you have something totally different in mind? May be I'm missing something.